### PR TITLE
Fixes incorrect warning of side-effects in #2393

### LINKED ISSUE #2393: wrong warning: side-effects from mutating the binding
Prepacking this program...
```js
(function() {
  function fn(arg) {
    if (arg) return undefined;
    let f = 42;
    return f;
  }
  __optimize(fn);
})();
```
gives the following warning:
```
side-effects from mutating the binding "f" at location: 4:12 - 4:4
```
Why?

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

Fixes #2393. I created the test as a React test because I couldn't see any way of making a serializer test fail if there was a warning without making many other tests fail or without adding another `// some config` to the test (which I believe we said we would not add anymore as there's already too many). React tests will always fail on side-effect warnings, so this takes better advantage of this.

### CODE REVIEW

File: src/realm.js [hunk 1]
```diff
@@ -1534,6 +1534,9 @@ export class Realm {
       if (root instanceof FunctionEnvironmentRecord && func === root.$FunctionObject) {
         return true;
       }
+      if (root === env.environmentRecord) {
```
Comment (NTillmann) [inline src/realm.js]: If the loop below would start at `env` and not `env.parent`, then you wouldn't need this special case. Or is it important that this check runs before the next check around created objects that can return `false`? I don't get the full picture of what is being checked here and the significance of ordering.

**Comment by trueadm** [comment]
@cblappert I'm not sure that overloading the `// expected` comment to have no error really makes a lot of sense to those unfamiliar with that particular comment.

File: src/realm.js (hunk 1 above)
Comment (trueadm) [reply src/realm.js]: That's a good point, I found we can leverage `env.destroyed` and use that loop and it simplifies the code quite a bit.

File: src/realm.js [hunk 2]
```diff
@@ -1534,12 +1534,8 @@ export class Realm {
       if (root instanceof FunctionEnvironmentRecord && func === root.$FunctionObject) {
         return true;
       }
-      if (this.createdObjectsTrackedForLeaks !== undefined && !this.createdObjectsTrackedForLeaks.has(func)) {
-        return false;
-      }
-      env = env.parent;
       while (env) {
-        if (env.environmentRecord === root) {
+        if (env.environmentRecord === root && !env.destroyed) {
```
[the remark this answers is not in the document]
Comment (trueadm) [reply src/realm.js]: It's checking that the environment record of the binding was that of the function we're in.

[the remark this answers is not in the document]
Comment (NTillmann) [reply src/realm.js]: And why exclude destroyed envs?

[the remark this answers is not in the document]
Comment (trueadm) [reply src/realm.js]: If they're excluded, it means we already executed that and destroyed the bindings in that scope so we don't check if that env anymore as it's bindings are no longer valid.

[the remark this answers is not in the document]
Comment (NTillmann) [reply src/realm.js]: But if the matching lexical environment has been destroyed, then `isDefinedInsidePureFn` will return `false`, then below we _would_ report side effect warnings? I am confused...

[the remark this answers is not in the document]
Comment (trueadm) [reply src/realm.js]: How can it be a pure function if the env has been destroyed? It should rightly report that there is a side-effect, as in the case of a pure function the env should never be destroyed.

[the remark this answers is not in the document]
Comment (NTillmann) [reply src/realm.js]: Okay, I think I understand. The destroyed-aspect is *the* key in determining whether this modified binding represents a side effect.

This simplified version of the function also passes all the public tests in your PR. Anything wrong with that? If it looks right, can we go with this version?

```js
    const isDefinedInsidePureFn = root => {
      let context = this.getRunningContext();
      let { lexicalEnvironment: env } = context;
      while (env) {
        if (env.environmentRecord === root) {
          // We can look at whether the lexical environment of the binding was destroyed to determine if it was defined outside the current pure running context.
          return !env.destroyed;
        }
        env = env.parent;
      }
      return false;
    };
```

**Comment by trueadm** [comment]
@NTillmann Sure, feel free to make the changes to this PR directly if you want.

File: src/realm.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (trueadm) [reply src/realm.js]: Updated PR with this code.

File: src/realm.js [hunk 3]
```diff
@@ -1528,19 +1528,12 @@ export class Realm {
...
-      invariant(func instanceof FunctionValue);
-      if (root instanceof FunctionEnvironmentRecord && func === root.$FunctionObject) {
-        return true;
-      }
-      if (this.createdObjectsTrackedForLeaks !== undefined && !this.createdObjectsTrackedForLeaks.has(func)) {
-        return false;
-      }
-      env = env.parent;
+      let { lexicalEnvironment: env } = context;
       while (env) {
```
Comment (hermanventer) [inline src/realm.js]: Not your fault, but while you are here, would it be possible to change this to "env !== null". We avoid implicit conversions to boolean whenever possible.

**Pull request closed without merging**